### ISSUE TITLE
multiple InferenceSession slowdown inference speed

### ISSUE THREAD

**Issue by jayden333** [microsoft/onnxruntime#10273]
**Describe the bug**
Multiple InferenceSession slowdown inference speed

**Urgency**
If there are particular important use cases blocked by this or strict project-related timelines, please share more information and dates. If there are no hard deadlines, please specify none.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux centos 7
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 1.10.0
- Python version: python3.6.8
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**To Reproduce**

use this script, when` run_single_sess()`, average time is 31ms, when `run_two_sess()`, average time is 46ms
model is download from https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/tiny-yolov2
cpu info: Intel(R) Xeon(R) Platinum 8255C CPU @ 2.50GHz, 10 cores for my docker container

```python
import time
import numpy as np
import onnxruntime as ort

model_path = "ft_local/tinyyolov2-8.onnx"

def run_two_sess():
    ort_sess1 = ort.InferenceSession(model_path)
    ort_sess2 = ort.InferenceSession(model_path)
    
    warm_up_count = 5
    for i in range(warm_up_count):
        x = np.random.rand(1, 3, 416, 416).astype("float32")
        ort_sess1.run(None, {ort_sess1._inputs_meta[0].name: x})
        x = np.random.rand(1, 3, 416, 416).astype("float32")
        ort_sess2.run(None, {ort_sess2._inputs_meta[0].name: x})
    
    count = 100
    start_total = time.time()
    for i in range(count):
        x = np.random.rand(1, 3, 416, 416).astype("float32")
        start = time.time()
        ort_sess1.run(None, {ort_sess1._inputs_meta[0].name: x})
        print(f"p1 consumed: {(time.time() - start)*1000}")
        x = np.random.rand(1, 3, 416, 416).astype("float32")
        start = time.time()
        ort_sess2.run(None, {ort_sess2._inputs_meta[0].name: x})
        print(f"p2 consumed: {(time.time() - start)*1000}")
    end_total = time.time()
    print(f"average time: {(end_total - start_total)*1000/count/2}")

def run_single_sess():
    ort_sess1 = ort.InferenceSession(model_path)
    
    warm_up_count = 5
    for i in range(warm_up_count):
        x = np.random.rand(1, 3, 416, 416).astype("float32")
        ort_sess1.run(None, {ort_sess1._inputs_meta[0].name: x})
    
    count = 100
    start_total = time.time()
    for i in range(count):
        x = np.random.rand(1, 3, 416, 416).astype("float32")
        start = time.time()
        ort_sess1.run(None, {ort_sess1._inputs_meta[0].name: x})
        print(f"p1 consumed: {(time.time() - start)*1000}")
    end_total = time.time()
    print(f"average time: {(end_total - start_total)*1000/count}")


if __name__ == "__main__":
    run_single_sess()
    # run_two_sess
```

**Expected behavior**
it should cost same time when I use two session, compared with single session

Appreciate for any help

**Comment by tianleiwu** [comment]
Is it expected since one session could use more threads (like one session uses 10 CPU cores, and two sessions could use 5 cores each).

A general way for two inference sessions is to have two processes, and each process have one inference session, and set affinity for each process to different CPU cores (In Lunux, use taskset or numctrl commands as in https://www.glennklockwood.com/hpc-howtos/process-affinity.html). This could be better than hosting two sessions in one process.

**Comment by pranavsharma** [comment]
You could also try sharing a single threadpool between 2 inference sessions using the global threadpool feature. By default each session has its own separate threadpool. Using 2 processes is better though.

**Comment by jayden333** [comment]
> Is it expected since one session could use more threads (like one session uses 10 CPU cores, and two sessions could use 5 cores each).
> 
> A general way for two inference sessions is to have two processes, and each process have one inference session, and set affinity for each process to different CPU cores (In Lunux, use taskset or numctrl commands as in https://www.glennklockwood.com/hpc-howtos/process-affinity.html). This could be better than hosting two sessions in one process.

@tianleiwu  thanks for your reply
I change my script, use `intra_op_num_threads` to limit thread pool size to 5:
```python
def run_single_sess():
    opt = ort.SessionOptions()
    opt.intra_op_num_threads = 5
    opt.inter_op_num_threads = 1
    ort_sess1 = ort.InferenceSession(model_path, opt)

def run_two_sess():
    opt1 = ort.SessionOptions()
    opt1.intra_op_num_threads = 5
    opt1.inter_op_num_threads = 1
    ort_sess1 = ort.InferenceSession(model_path, opt1)
    opt2 = ort.SessionOptions()
    opt2.intra_op_num_threads = 5
    opt2.inter_op_num_threads = 1
    ort_sess2 = ort.InferenceSession(model_path, opt2)
```
the result is still strange, when two session exist, the `run()` speed slow down

**Comment by jayden333** [comment]
> 

I can't find the API in python for sharing a single threadpool between sessions, could you help show me a demo? thanks

**Comment by jayden333** [comment]
> You could also try sharing a single threadpool between 2 inference sessions using the global threadpool feature. By default each session has its own separate threadpool. Using 2 processes is better though.

@pranavsharma  I can't find the API in python for sharing a single threadpool between sessions, could you help show me a demo? thanks

**Comment by pranavsharma** [comment]
> > You could also try sharing a single threadpool between 2 inference sessions using the global threadpool feature. By default each session has its own separate threadpool. Using 2 processes is better though.
> 
> @pranavsharma I can't find the API in python for sharing a single threadpool between sessions, could you help show me a demo? thanks

Sorry, this feature is not available in the Python API.

**Comment by jayden333** [comment]
> 



> Is it expected since one session could use more threads (like one session uses 10 CPU cores, and two sessions could use 5 cores each).
> 
> A general way for two inference sessions is to have two processes, and each process have one inference session, and set affinity for each process to different CPU cores (In Lunux, use taskset or numctrl commands as in https://www.glennklockwood.com/hpc-howtos/process-affinity.html). This could be better than hosting two sessions in one process.

@tianleiwu I tried use two process and each use different CPU cores(by taskset), but still same result

**Comment by emhagman** [comment]
> Sorry, this feature is not available in the Python API.

Is this still the case?

**Comment by pranavsharma** [comment]
Yeah, you can now create multiple sessions with limited number of threads and pin them to separate cores using the 'intra_op_thread_affinities' config. This is not really the same as sharing of the global threadpool between sessions but it'll do it for you and might be better due to thread pinning.

**Comment by emhagman** [comment]
> Yeah, you can now create multiple sessions with limited number of threads and pin them to separate cores using the 'intra_op_thread_affinities' config. This is not really the same as sharing of the global threadpool between sessions but it'll do it for you and might be better due to thread pinning.

Thank you for responding so quickly! Understood. So, in theory, there still can be resource contention here, though? I am using on `rt.SessionOptions` object for all of my sessions (16 in my case). So if I use the affinity, will they all use the same worker pool? In my case, would I want to lock each model onto different logical cores so they aren't competing?

**Comment by pranavsharma** [comment]
Yeah, you want to pin the threads from each session to separate cores to avoid any contention.
